src/dsl/cb.rs

```diff
@@ -92,28 +115,50 @@ pub fn or<
 
     for constraint in inputs.into_iter() {
         let constraint = constraint.into();
-        annotations.push(constraint.annotation);
-        exprs.push(constraint.expr);
+        match constraint.typing {
+            Typing::Boolean | Typing::Unknown => {
+                annotations.push(constraint.annotation);
+                exprs.push(constraint.expr);
+            }
+            Typing::AntiBooly => panic!("Expected Boolean or Unknown constraint, got AntiBooly (constraint: {})", constraint.annotation),
+        }
     }
 
     let result = not(and(exprs.into_iter().map(not)));
 
     Constraint {
         annotation: format!("({})", annotations.join(" OR ")),
         expr: result.expr,
+        typing: Typing::Boolean,
     }
 }
 
 /// Takes two expressions and returns a new expression representing the logical XOR of the input
 /// expressions.
-pub fn xor<F: From<u64> + Clone, LHS: Into<Expr<F>>, RHS: Into<Expr<F>>>(
+pub fn xor<F: From<u64> + Clone, LHS: Into<Constraint<F>>, RHS: Into<Constraint<F>>>(
     lhs: LHS,
     rhs: RHS,
-) -> Expr<F> {
-    let lhs = lhs.into();
-    let rhs = rhs.into();
+) -> Constraint<F> {
+    let mut annotations: Vec<String> = vec![];
+    let expr: Expr<F>;
 
-    lhs.clone() + rhs.clone() - 2u64.expr() * lhs * rhs
+    let lhs: Constraint<F> = lhs.into();
+    let rhs: Constraint<F> = rhs.into();
+
+    match (lhs.typing, rhs.typing) {
```

Comment:
this is not important but you coud do:

`let expr = match ...` and to don't put semicolon after the expression in the last line